btrfs graphdriver readonly subvolumes capability

nikitych opened the pull request:
<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/moby/moby/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

For additional information on our contributing process, read our contributing
guide https://docs.docker.com/opensource/code/

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Please provide the following information:
-->

**- What I did**
Fixes #35630 . btrfs graphdriver create readonly subvolumes by default and mark them writable when _Get_ is calling.

**- How I did it**
I add _subvolSetPropRO_ function to btrfs graphdriver which could set readonly property to subvolume, and call it at _Get_, _Put_ or _subvolDelete_.
Also rework _subvolDelete_ to increase performance and handle some corner cases.

**- How to verify it**
Intermediate layers in `/var/lib/docker/btrfs/subvolumes/` are read-only. Try to write something and you'll see.

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->
btrfs graphdriver keeps subvolumes in read-only state when possible

thaJeztah commented:
/cc @cyphar as well

kolyshkin commented:
First commit looks OK. Second (the one that fixes `layer_store.registerWithDescriptor` is not quite comprehensible. It needs a description in the changelog explaining what is the issue it fixes, and maybe some code cleanup with respect to using `layer.cacheID`, `layerCacheID` and `writableLayerCacheID` variables.

kolyshkin commented:
OK, it seems the semantics of `CreateReadWrite()`, as introduced by commit ef5bfad3210a9e, implied something different. I wonder how the changes in `layer_store.registerWithDescriptor()` will affect other graph drivers...

nikitych commented:
Thanks for point to `CreateReadWrite` origin. Looks like fix in `registerWithDescriptor` broke badly the windows part. Haven't at hand win host to debug.

nikitych committed:
Change btrfs graphdriver to keep subvolumes in readonly state until used.

Also remove reference to deprecated IOCTL BTRFS_IOC_SUBVOL_CREATE

Signed-off-by: Nikita Gerasimov <nikita.gerasimov@oracle.com>

nikitych committed:
Rework subvolDelete to a more efficient algorithm.

As a side effect change openDir to handle correctly a long paths (>=PATH_MAX).

Signed-off-by: Nikita Gerasimov <nikita.gerasimov@oracle.com>

thaJeztah commented:
@nikitych any reason you closed the PR?

[fef616327d13b7faf19d9d4b79c903aee91b63a2.patch.txt](https://github.com/moby/moby/files/1719948/fef616327d13b7faf19d9d4b79c903aee91b63a2.patch.txt)
[4150e9cb1d89d4b20597f2619dee4d711ae9b484.patch.txt](https://github.com/moby/moby/files/1719949/4150e9cb1d89d4b20597f2619dee4d711ae9b484.patch.txt)

nikitych commented:
@thaJeztah Found some bugs, need more tests to understand if  4150e9cb1d89d4b20597f2619dee4d711ae9b484.patch.txt general applicable solution.
